### PR TITLE
[ColossalChat] Update RLHF V2

### PR DESCRIPTION
**Pull request by YeAnbang**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [ ] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [ ] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

**Comment by TongLi3701** [comment]
Hi Anbang, is this ready to review? Should I review https://github.com/hpcaitech/ColossalAI/pull/5148 this first?

**Comment by YeAnbang** [comment]
Is ready for review now. Still working on fixing the following bug:

- All script hang during booster.save_model with Gemini.

**Review by TongLi3701 — COMMENTED** [review]
Thanks Anbang. Overall looks great.

Some minor comments below. 

In additional, some of the links are out of date, please check. 

Also, for PPO and DPO part, as there are multiple models are running (inference and training), please add a table for minimal resource needed, e.g., 7B model

File: applications/ColossalChat/benchmarks/benchmark_ppo.py [hunk 1]
```diff
@@ -0,0 +1,493 @@
...
+    else:
+        raise ValueError(f"Unknown plugin {args.plugin}")
+
+    if args.plugin != "3d":
+        custom_plugin = plugin
+
+    actor_booster = Booster(plugin=plugin)
+    ref_booster = Booster(plugin=plugin)
+    rm_booster = Booster(plugin=custom_plugin)
+    critic_booster = Booster(plugin=custom_plugin)
```
[the remark this answers is not in the document]
Comment (YeAnbang) [reply applications/ColossalChat/benchmarks/benchmark_ppo.py:294]: Our system people said so

File: applications/ColossalChat/examples/README.md [hunk 2]
```diff
@@ -0,0 +1,558 @@
...
+Finishing the above steps, you have converted the raw conversation to the designated chat format and tokenized the formatted conversation, calculate input_ids, labels, attention_masks and buffer those into binary dataset files under "$SAVE_DIR/arrow/part-XXXX" folders.
+
+For now, ColossalChat only support chat models whose chat template is in the form of,
+```json
+<some additional tokens>[system message]<human_line_start>[human line]<human_line_end><assistant_line_start>[assistant line]<assistant_line_end>...[assistant line]<assistant_line_end><some additional tokens>
+```
+
+For example, our Colossal-LLaMA-2 format looks like,
+```
+<s> A chat between a curious human and an artificial intelligence assistant. The assistant gives helpful, detailed, and polite answers to the human's questions.
```
[the remark this answers is not in the document]
Comment (YeAnbang) [reply applications/ColossalChat/examples/README.md:362]: Double-checked, there is no </s> at the end of the system prompt. There is no error

File: applications/ColossalChat/examples/data_preparation_scripts/prepare_sft_dataset.py [hunk 3]
```diff
@@ -0,0 +1,179 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""
+Prepare sft dataset for finetuning
+"""
+
```
[the remark this answers is not in the document]
Comment (YeAnbang) [reply applications/ColossalChat/examples/data_preparation_scripts/prepare_sft_dataset.py:6]: moved all data preparation code into one script

File: applications/ColossalChat/coati/utils/ckpt_io.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+    # booster.save_optimizer(optimizer, os.path.join(save_dir, "optimizer"), shard=True)
+    # booster.save_lr_scheduler(lr_scheduler, os.path.join(save_dir, "lr_scheduler"))
+    # running_states = {
+    #     "epoch": epoch,
+    #     "step": step,
+    #     "sample_start_index": step * batch_size,
+    # }
+    # if coordinator.is_master():
+    #     save_json(running_states, os.path.join(save_dir, "running_states.json"))
```
[the remark this answers is not in the document]
Comment (YeAnbang) [reply applications/ColossalChat/coati/utils/ckpt_io.py:69]: resolved

File: applications/ColossalChat/examples/training_scripts/train_ppo.py [hunk 5]
```diff
@@ -0,0 +1,482 @@
...
+    if args.plugin != "3d":
+        custom_plugin = plugin
+
+    actor_booster = Booster(plugin=plugin)
+    ref_booster = Booster(plugin=plugin)
+    rm_booster = Booster(plugin=custom_plugin)
+    critic_booster = Booster(plugin=custom_plugin)
+
+    default_dtype = torch.float16 if args.mixed_precision == "fp16" else torch.bfloat16
+    torch.set_default_dtype(default_dtype)
```
[the remark this answers is not in the document]
Comment (YeAnbang) [reply applications/ColossalChat/examples/training_scripts/train_ppo.py:280]: Not sure if this is a duplication. https://github.com/hpcaitech/ColossalAI/blob/bce9499ed33f7a8359bbb568c7ee18d72e8aa731/applications/Colossal-LLaMA-2/train_sft.py#L260

**Comment by YeAnbang** [comment]
# Local CI Test Results
[ci.pdf](https://github.com/hpcaitech/ColossalAI/files/14138681/ci.pdf)

**Comment by YeAnbang** [comment]
all tests passed locally.

image

**Comment by YeAnbang** [comment]
# Experiment Report

## SFT
image

## DPO
img_v3_029a_1201f32f-cdba-40e7-923e-d8ecc0f2d66g

## PPO
img_v3_029a_a70dba2a-7d80-4ee1-935f-67f5a47d17fg
5FrWy9zwTr

**Review by TongLi3701 — APPROVED** [review]
Thanks Anbang. Overall looks ok.

**Pull request merged**